File: chairlift/window.py
# window.py
#
# Copyright 2023 mirkobrombin
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundationat version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import threading

from gi.repository import Gtk, Adw, GLib, Gio, Gdk

import chairlift.core.backend as backend

from chairlift.dialog import ChairLiftDialog

_ = __builtins__["_"]

@Gtk.Template(resource_path="/org/frostyard/ChairLift/gtk/window.ui")
class ChairLiftWindow(Adw.ApplicationWindow):
    __gtype_name__ = "ChairLiftWindow"

    split_view = Gtk.Template.Child()
    sidebar_list = Gtk.Template.Child()
    content_stack = Gtk.Template.Child()
    toasts = Gtk.Template.Child()
    style_manager = Adw.StyleManager().get_default()

    pages = {}

    def __init__(self, moduledir: str,  **kwargs):
        super().__init__(**kwargs)

        self.moduledir = moduledir

        self.__build_ui()
        self.__setup_actions()

        backend.subscribe_errors(self.__error_received)


    def __error_received(self, script_name: str, command: list[str], id: int):
        GLib.idle_add(self.__error_toast, _("Execution failed: ") + script_name, id)

    def add_toast(self, toast: any):

        self.toasts.add_toast(toast)


    def __error_toast(self, message: str, id: int):
        toast = Adw.Toast.new(message)
        toast.props.timeout = 0
        toast.props.button_label = _("Details")
        toast.connect("button-clicked", self.__error_toast_clicked, id)
        self.toasts.add_toast(toast)

    def __error_toast_clicked(self, widget, id: int):
        message = backend.errors[id]
        dialog = ChairLiftDialog(self, _("Error log"), message)
        dialog.present()




    def __build_ui(self):
        from chairlift.views.user_home import ChairLiftUserHome

        self.__view_home = ChairLiftUserHome(self)

        # Setup sidebar navigation items
        nav_items = [
            {"name": "applications", "title": _("Applications"), "icon": "application-x-executable-symbolic"},
            {"name": "maintenance", "title": _("Maintenance"), "icon": "emblem-system-symbolic"},
            {"name": "updates", "title": _("Updates"), "icon": "software-update-available-symbolic"},
            {"name": "system", "title": _("System"), "icon": "computer-symbolic"},
            {"name": "help", "title": _("Help"), "icon": "help-browser-symbolic"},
        ]

        for item in nav_items:
            row = Adw.ActionRow()
            row.set_title(item["title"])
            row.set_icon_name(item["icon"])
            row.set_activatable(True)
            row.page_name = item["name"]
            self.sidebar_list.append(row)
            
            # Get page from view_home
            page = self.__view_home.get_page(item["name"])
            if page:
                self.pages[item["name"]] = page
                self.content_stack.add_child(page)
        
        # Connect sidebar selection
        self.sidebar_list.connect("row-activated", self.__on_sidebar_row_activated)
        
        # Select first item by default
        self.sidebar_list.select_row(self.sidebar_list.get_row_at_index(0))
        if nav_items:
            first_page = self.pages.get(nav_items[0]["name"])
            if first_page:
                self.content_stack.set_visible_child(first_page)
        
        # Enable window controls
        self.set_deletable(True)
    
    def __on_sidebar_row_activated(self, listbox, row):
        """Handle sidebar navigation"""
        page_name = getattr(row, 'page_name', None)
        if page_name and page_name in self.pages:
            self.content_stack.set_visible_child(self.pages[page_name])
            # Show content page in narrow mode
            self.split_view.set_show_content(True)
    
    def __setup_actions(self):
        """Setup window actions"""
        # Keyboard shortcuts action
        shortcuts_action = Gio.SimpleAction.new("show-shortcuts", None)
        shortcuts_action.connect("activate", self.__on_show_shortcuts)
        self.add_action(shortcuts_action)
        
        # Help action
        help_action = Gio.SimpleAction.new("show-help", None)
        help_action.connect("activate", self.__on_show_help)
        self.add_action(help_action)
        
        # About action
        about_action = Gio.SimpleAction.new("show-about", None)
        about_action.connect("activate", self.__on_show_about)
        self.add_action(about_action)
        
        # Navigation actions
        for i, page_name in enumerate(["applications", "maintenance", "updates", "system", "help"], start=1):
            action = Gio.SimpleAction.new(f"navigate-{page_name}", None)
            action.connect("activate", self.__on_navigate_to_page, page_name)
            self.add_action(action)
    
    def __on_show_shortcuts(self, action, param):
        """Show keyboard shortcuts window"""
        builder = Gtk.Builder.new_from_string("""
<?xml version="1.0" encoding="UTF-8"?>
<interface>
  <object class="GtkShortcutsWindow" id="shortcuts_window">
    <property name="modal">true</property>
    <child>
      <object class="GtkShortcutsSection">
        <property name="section-name">shortcuts</property>
        <child>
          <object class="GtkShortcutsGroup">
            <property name="title" translatable="yes">General</property>
            <child>
              <object class="GtkShortcutsShortcut">
                <property name="title" translatable="yes">Show Keyboard Shortcuts</property>
                <property name="accelerator">&lt;Primary&gt;question</property>
              </object>
            </child>
            <child>
              <object class="GtkShortcutsShortcut">
                <property name="title" translatable="yes">Quit</property>
                <property name="accelerator">&lt;Primary&gt;q</property>
              </object>
            </child>
          </object>
        </child>
        <child>
          <object class="GtkShortcutsGroup">
            <property name="title" translatable="yes">Navigation</property>
            <child>
              <object class="GtkShortcutsShortcut">
                <property name="title" translatable="yes">Applications</property>
                <property name="accelerator">&lt;Alt&gt;1</property>
              </object>
            </child>
            <child>
              <object class="GtkShortcutsShortcut">
                <property name="title" translatable="yes">Maintenance</property>
                <property name="accelerator">&lt;Alt&gt;2</property>
              </object>
            </child>
            <child>
              <object class="GtkShortcutsShortcut">
                <property name="title" translatable="yes">Updates</property>
                <property name="accelerator">&lt;Alt&gt;3</property>
              </object>
            </child>
            <child>
              <object class="GtkShortcutsShortcut">
                <property name="title" translatable="yes">System</property>
                <property name="accelerator">&lt;Alt&gt;4</property>
              </object>
            </child>
            <child>
              <object class="GtkShortcutsShortcut">
                <property name="title" translatable="yes">Help</property>
                <property name="accelerator">&lt;Alt&gt;5</property>
              </object>
            </child>
          </object>
        </child>
      </object>
    </child>
  </object>
</interface>
""", -1)
        shortcuts_window = builder.get_object("shortcuts_window")
        shortcuts_window.set_transient_for(self)
        shortcuts_window.present()
    
    def __on_show_help(self, action, param):
        """Show help documentation"""
        Gtk.show_uri(self, "https://github.com/frostyard/chairlift", Gdk.CURRENT_TIME)
    
    def __on_show_about(self, action, param):
        """Show about dialog"""
        about = Adw.AboutDialog.new()
        about.set_application_name("ChairLift")
        about.set_application_icon("org.frostyard.ChairLift")
        about.set_version("VTESTING")
        about.set_developer_name("Brian Ketelsen")
        about.set_license_type(Gtk.License.GPL_3_0)
        about.set_comments(_("System management and configuration tool"))
        about.set_website("https://github.com/frostyard/chairlift")
        about.set_issue_url("https://github.com/frostyard/chairlift/issues")
        about.set_developers([
            "Brian Ketelsen https://github.com/bketelsen",
        ])
        about.set_copyright("© 2025 FrostYard")
        about.present(self)
    
    def __on_navigate_to_page(self, action, param, page_name):
        """Navigate to a specific page using keyboard shortcut"""
        if page_name in self.pages:
            self.content_stack.set_visible_child(self.pages[page_name])
            # Select the corresponding row in the sidebar
            for i, row in enumerate([self.sidebar_list.get_row_at_index(i) for i in range(5)]):
                if hasattr(row, 'page_name') and row.page_name == page_name:
                    self.sidebar_list.select_row(row)
                    break
            # Show content page in narrow mode
            self.split_view.set_show_content(True)

File: internal/window/window.go
// Package window provides the main application window
package window

import (
	"fmt"

	"github.com/frostyard/chairlift/internal/config"
	"github.com/frostyard/chairlift/internal/views"

	"github.com/jwijenbergh/puregotk/v4/adw"
	"github.com/jwijenbergh/puregotk/v4/gio"
	"github.com/jwijenbergh/puregotk/v4/gtk"
)

// Window represents the main application window
type Window struct {
	*adw.ApplicationWindow

	splitView    *adw.NavigationSplitView
	sidebarList  *gtk.ListBox
	contentStack *gtk.Stack
	contentPage  *adw.NavigationPage // Content navigation page for dynamic title
	toasts       *adw.ToastOverlay

	pages       map[string]*adw.ToolbarView
	navRows     map[string]*adw.ActionRow // Store references to nav rows for badges
	config      *config.Config
	views       *views.UserHome
	updateBadge *gtk.Button // Badge for updates count
}

// NavItem represents a navigation item in the sidebar
type NavItem struct {
	Name  string
	Title string
	Icon  string
}

// navItems defines the sidebar navigation structure
var navItems = []NavItem{
	{Name: "applications", Title: "Applications", Icon: "application-x-executable-symbolic"},
	{Name: "maintenance", Title: "Maintenance", Icon: "emblem-system-symbolic"},
	{Name: "updates", Title: "Updates", Icon: "software-update-available-symbolic"},
	{Name: "system", Title: "System", Icon: "computer-symbolic"},
	{Name: "extensions", Title: "Extensions", Icon: "application-x-addon-symbolic"},
	{Name: "help", Title: "Help", Icon: "help-browser-symbolic"},
}

// New creates a new main window
func New(app *adw.Application) *Window {
	w := &Window{
		ApplicationWindow: adw.NewApplicationWindow(&app.Application),
		pages:             make(map[string]*adw.ToolbarView),
		navRows:           make(map[string]*adw.ActionRow),
		config:            config.Load(),
	}

	w.SetDefaultSize(900, 700)
	w.SetTitle("ChairLift")

	w.buildUI()
	w.setupActions()

	return w
}

// buildUI constructs the window UI
func (w *Window) buildUI() {
	// Create views manager
	w.views = views.New(w.config, w)

	// Create the navigation split view
	w.splitView = adw.NewNavigationSplitView()

	// Create sidebar
	sidebarPage := w.buildSidebar()
	w.splitView.SetSidebar(sidebarPage)

	// Create content area
	contentPage := w.buildContentArea()
	w.splitView.SetContent(contentPage)

	// Create toast overlay for notifications
	w.toasts = adw.NewToastOverlay()
	w.toasts.SetChild(&w.splitView.Widget)

	// Set window content
	w.SetContent(&w.toasts.Widget)
}

// buildSidebar creates the sidebar navigation
func (w *Window) buildSidebar() *adw.NavigationPage {
	// Create toolbar view for sidebar
	toolbarView := adw.NewToolbarView()

	// Add header bar with menu button
	headerBar := adw.NewHeaderBar()
	headerBar.SetShowEndTitleButtons(false)

	// Create hamburger menu button
	menuButton := w.buildMenuButton()
	headerBar.PackEnd(&menuButton.Widget)

	toolbarView.AddTopBar(&headerBar.Widget)

	// Create scrolled window for the list
	scrolled := gtk.NewScrolledWindow()
	scrolled.SetPolicy(gtk.PolicyNeverValue, gtk.PolicyAutomaticValue)
	scrolled.SetVexpand(true)

	// Create list box for navigation
	w.sidebarList = gtk.NewListBox()
	w.sidebarList.SetSelectionMode(gtk.SelectionSingleValue)
	w.sidebarList.AddCssClass("navigation-sidebar")

	// Add navigation items
	for _, item := range navItems {
		row := w.createNavRow(item)
		w.sidebarList.Append(&row.Widget)
	}

	// Connect row activation
	rowActivatedCb := func(listbox gtk.ListBox, rowPtr uintptr) {
		// Convert uintptr to ListBoxRow
		row := gtk.ListBoxRowNewFromInternalPtr(rowPtr)
		w.onSidebarRowActivated(*row)
	}
	w.sidebarList.ConnectRowActivated(&rowActivatedCb)

	scrolled.SetChild(&w.sidebarList.Widget)
	toolbarView.SetContent(&scrolled.Widget)

	// Create navigation page
	navPage := adw.NewNavigationPage(&toolbarView.Widget, "ChairLift")

	return navPage
}

// createNavRow creates a navigation row for the sidebar
func (w *Window) createNavRow(item NavItem) *adw.ActionRow {
	row := adw.NewActionRow()
	row.SetTitle(item.Title)
	row.SetActivatable(true)

	// Add icon
	icon := gtk.NewImageFromIconName(item.Icon)
	row.AddPrefix(&icon.Widget)

	// Add badge for updates row (hidden by default)
	if item.Name == "updates" {
		w.updateBadge = gtk.NewButton()
		w.updateBadge.AddCssClass("circular")
		w.updateBadge.AddCssClass("warning")
		w.updateBadge.SetVisible(false)
		row.AddSuffix(&w.updateBadge.Widget)
	}

	// Store the page name in the row (using SetName for identification)
	row.SetName(item.Name)

	// Store reference to the row
	w.navRows[item.Name] = row

	return row
}

// buildContentArea creates the content stack
func (w *Window) buildContentArea() *adw.NavigationPage {
	// Create stack for content pages
	w.contentStack = gtk.NewStack()
	w.contentStack.SetTransitionType(gtk.StackTransitionTypeCrossfadeValue)

	// Add pages to the stack
	for _, item := range navItems {
		page := w.views.GetPage(item.Name)
		if page != nil {
			w.pages[item.Name] = page
			w.contentStack.AddNamed(&page.Widget, item.Name)
		}
	}

	// Create navigation page with initial title from first nav item
	initialTitle := "Content"
	if len(navItems) > 0 {
		initialTitle = navItems[0].Title
	}
	w.contentPage = adw.NewNavigationPage(&w.contentStack.Widget, initialTitle)

	// Select first item by default
	if len(navItems) > 0 {
		firstRow := w.sidebarList.GetRowAtIndex(0)
		if firstRow != nil {
			w.sidebarList.SelectRow(firstRow)
			w.contentStack.SetVisibleChildName(navItems[0].Name)
		}
	}

	return w.contentPage
}

// onSidebarRowActivated handles sidebar row activation
func (w *Window) onSidebarRowActivated(row gtk.ListBoxRow) {
	// Get the ActionRow from the ListBoxRow
	widget := row.GetChild()
	if widget == nil {
		return
	}

	// Get the name we stored
	name := row.GetName()
	if name == "" {
		return
	}

	// Switch to the corresponding page
	if _, ok := w.pages[name]; ok {
		w.contentStack.SetVisibleChildName(name)
		w.splitView.SetShowContent(true)

		// Update the content page title
		for _, item := range navItems {
			if item.Name == name {
				w.contentPage.SetTitle(item.Title)
				break
			}
		}
	}
}

// buildMenuButton creates the hamburger menu button
func (w *Window) buildMenuButton() *gtk.MenuButton {
	// Create menu model
	menu := gio.NewMenu()

	// Add menu items
	menu.Append("Keyboard Shortcuts", "win.show-shortcuts")
	menu.Append("About ChairLift", "win.show-about")

	// Create menu button
	menuButton := gtk.NewMenuButton()
	menuButton.SetIconName("open-menu-symbolic")
	menuButton.SetMenuModel(&menu.MenuModel)
	menuButton.SetTooltipText("Main Menu")

	return menuButton
}

// setupActions sets up window actions
func (w *Window) setupActions() {
	// Show shortcuts action
	shortcutsAction := gio.NewSimpleAction("show-shortcuts", nil)
	shortcutsActivateCb := func(action gio.SimpleAction, param uintptr) {
		w.onShowShortcuts()
	}
	shortcutsAction.ConnectActivate(&shortcutsActivateCb)
	w.AddAction(shortcutsAction)

	// About action
	aboutAction := gio.NewSimpleAction("show-about", nil)
	aboutActivateCb := func(action gio.SimpleAction, param uintptr) {
		w.onShowAbout()
	}
	aboutAction.ConnectActivate(&aboutActivateCb)
	w.AddAction(aboutAction)

	// Navigation actions
	for _, item := range navItems {
		itemName := item.Name // Capture for closure
		action := gio.NewSimpleAction("navigate-"+itemName, nil)
		navActivateCb := func(action gio.SimpleAction, param uintptr) {
			w.navigateToPage(itemName)
		}
		action.ConnectActivate(&navActivateCb)
		w.AddAction(action)
	}
}

// navigateToPage navigates to a specific page
func (w *Window) navigateToPage(pageName string) {
	if _, ok := w.pages[pageName]; ok {
		w.contentStack.SetVisibleChildName(pageName)

		// Select the corresponding row and update title
		for i, item := range navItems {
			if item.Name == pageName {
				row := w.sidebarList.GetRowAtIndex(i)
				if row != nil {
					w.sidebarList.SelectRow(row)
				}
				w.contentPage.SetTitle(item.Title)
				break
			}
		}
	}
}

// onShowShortcuts shows the keyboard shortcuts window
func (w *Window) onShowShortcuts() {
	// Create a dialog to show shortcuts since GtkShortcutsWindow isn't available in puregotk
	dialog := adw.NewWindow()
	dialog.SetTransientFor(&w.Window)
	dialog.SetModal(true)
	dialog.SetTitle("Keyboard Shortcuts")
	dialog.SetDefaultSize(400, 450)

	// Create toolbar view
	toolbarView := adw.NewToolbarView()

	// Add header bar
	headerBar := adw.NewHeaderBar()
	toolbarView.AddTopBar(&headerBar.Widget)

	// Create scrolled window
	scrolled := gtk.NewScrolledWindow()
	scrolled.SetPolicy(gtk.PolicyNeverValue, gtk.PolicyAutomaticValue)
	scrolled.SetVexpand(true)

	// Create main box
	mainBox := gtk.NewBox(gtk.OrientationVerticalValue, 0)
	mainBox.SetMarginTop(12)
	mainBox.SetMarginBottom(12)
	mainBox.SetMarginStart(12)
	mainBox.SetMarginEnd(12)

	// Create clamp for content width
	clamp := adw.NewClamp()
	clamp.SetMaximumSize(400)

	// Navigation shortcuts group
	navGroup := adw.NewPreferencesGroup()
	navGroup.SetTitle("Navigation")

	navShortcuts := []struct {
		accel string
		title string
	}{
		{"Alt+1", "Go to Applications"},
		{"Alt+2", "Go to Maintenance"},
		{"Alt+3", "Go to Updates"},
		{"Alt+4", "Go to System"},
		{"Alt+5", "Go to Help"},
	}

	for _, s := range navShortcuts {
		row := adw.NewActionRow()
		row.SetTitle(s.title)

		label := gtk.NewLabel(s.accel)
		label.AddCssClass("dim-label")
		row.AddSuffix(&label.Widget)

		navGroup.Add(&row.Widget)
	}

	mainBox.Append(&navGroup.Widget)

	// General shortcuts group
	generalGroup := adw.NewPreferencesGroup()
	generalGroup.SetTitle("General")

	generalShortcuts := []struct {
		accel string
		title string
	}{
		{"Ctrl+?", "Keyboard Shortcuts"},
		{"Ctrl+Q", "Quit"},
		{"F1", "Help"},
	}

	for _, s := range generalShortcuts {
		row := adw.NewActionRow()
		row.SetTitle(s.title)

		label := gtk.NewLabel(s.accel)
		label.AddCssClass("dim-label")
		row.AddSuffix(&label.Widget)

		generalGroup.Add(&row.Widget)
	}

	mainBox.Append(&generalGroup.Widget)

	clamp.SetChild(&mainBox.Widget)
	scrolled.SetChild(&clamp.Widget)
	toolbarView.SetContent(&scrolled.Widget)

	dialog.SetContent(&toolbarView.Widget)
	dialog.Present()
}

// onShowAbout shows the about dialog
func (w *Window) onShowAbout() {
	about := adw.NewAboutWindow()
	about.SetTransientFor(&w.Window)
	about.SetApplicationName("ChairLift")
	about.SetApplicationIcon("org.frostyard.ChairLift")
	about.SetVersion("1.0.0")
	about.SetDeveloperName("Frostyard")
	about.SetWebsite("https://github.com/frostyard/chairlift")
	about.SetIssueUrl("https://github.com/frostyard/chairlift/issues")
	about.SetLicenseType(gtk.LicenseGpl30Value)
	about.SetCopyright("© 2024-2026 Frostyard")
	about.SetDevelopers([]string{"Brian Ketelsen", "ChairLift Contributors"})
	about.Present()
}

// AddToast adds a toast notification
func (w *Window) AddToast(toast *adw.Toast) {
	w.toasts.AddToast(toast)
}

// ShowToast shows a simple toast message
func (w *Window) ShowToast(message string) {
	toast := adw.NewToast(message)
	toast.SetTimeout(3)
	w.AddToast(toast)
}

// ShowErrorToast shows an error toast
func (w *Window) ShowErrorToast(message string) {
	toast := adw.NewToast(message)
	toast.SetTimeout(0) // Persist until dismissed
	w.AddToast(toast)
}

// SetUpdateBadge updates the badge on the Updates navigation row
func (w *Window) SetUpdateBadge(count int) {
	if w.updateBadge == nil {
		return
	}

	if count > 0 {
		w.updateBadge.SetLabel(fmt.Sprintf("%d", count))
		w.updateBadge.SetVisible(true)
	} else {
		w.updateBadge.SetVisible(false)
	}
}
